autocmd: RecordingEnter, RecordingLeave  ("macro" events)

vigoux opened the pull request:
Add some autocmds to determine when the user is recording macros.

TODO:
- [x] Add `RecordingStart` and `RecordingStop` autocmds
- [x] Add `reg_recorded()`, to query what was the latest recorded register
- [x] Provide a default mapping using these features

Fixes partially #15404, cc @clason @justinmk

vigoux committed:
feat(autocmd): add Recording autocmds

vigoux committed:
feat(eval): add reg_recorded()

This function is used the get the last recorded register.

vigoux commented on runtime/doc/eval.txt:
```diff
@@ -6942,6 +6943,11 @@ reg_recording()						*reg_recording()*
 		Returns the single letter name of the register being recorded.
 		Returns an empty string string when not recording.  See |q|.
 
+reg_recording()						*reg_recording()*
```
Indeed, copypasta error hehe.

vigoux replied:
Should be fixed

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -4660,7 +4660,12 @@ static void nv_exmode(cmdarg_T *cap)
   if (VIsual_active) {
     vim_beep(BO_EX);
   } else if (!checkclearop(cap->oap)) {
-    do_exmode();
```
Maybe `Q` should be allowed in Visual mode, since both `q` and `@` are allowed.

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -4660,7 +4660,12 @@ static void nv_exmode(cmdarg_T *cap)
   if (VIsual_active) {
```
Maybe `Q` should be allowed in Visual mode, since both `q` and `@` are allowed.

vigoux replied:
Indeed, it would be great to have that too, working on it right now, and adding tests for it.

vigoux replied:
Both issues should be addressed now.

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -812,6 +812,14 @@ RemoteReply			When a reply from a Vim that functions as
 				the reply should be read with |remote_read()|
 				to consume it.
+							*RecordingStart*
+RecordingStart			When a macro starts to be recorded.
```
no precedent for "Start"/"Stop" in event names. I guess "Enter"/"Leave" make sense?

vigoux replied:
That's what I thought too, but it felt a bit weird... But if that's what you want, I'll do t

justinmk replied:
I'm ok with either.

vigoux replied:
Went for `Enter` and `Leave` for consistence.

justinmk commented on src/nvim/normal.c:
```diff
@@ -4652,15 +4652,18 @@ static void nv_zet(cmdarg_T *cap)
+
+  while (cap->count1-- && !got_int) {
+    if (do_execreg(reg_recorded, false, false, false) == false) {
+      beep_flush();
```
nv_at uses clearopbeep() here, is it not used here intentionally?

vigoux replied:
I guess it's an error from me.

justinmk commented on test/functional/eval/reg_spec.lua:
```diff
@@ -0,0 +1,16 @@
+local helpers = require('test.functional.helpers')(after_each)
```
let's put all of the test in macro_spec.lua. there is no reason for them to be separate really.

vigoux commented on src/nvim/ops.c:
```diff
@@ -909,35 +909,35 @@ int do_record(int c)
+    // Get the recorded key hits.  K_SPECIAL and CSI will be escaped, this
+    // needs to be removed again to put it in a register.  exec_reg then
+    // adds the escaping back later.
+    reg_recorded = reg_recording;
```
Then I follow your opinion !

vigoux replied:
I find that having `reg_recorded` invalid on `RecordingLeave` a bit confusing to be honest, are you sure it is fine ?

justinmk replied:
how is it invalid, if we define RecordingLeave as happening "just before" or "at the edge"?

vigoux replied:
That sounds a bit weird to me, that on `RecordingLeave`, one cannot use  the value of `reg_recorded` as it is not what it is expected to be.
_AND_ I was wrong and at `RecordingLeave` `reg_recording` will actually return an empty string.

justinmk replied:
> at `RecordingLeave` `reg_recording` will actually return an empty string.

but it doesn't have to. the point is for the even to be useful, why should we force extra bookkeeping for plugin authors?

vigoux replied:
As you want then...

vigoux committed:
docs(reg_recorded): add links to relevant docs

vigoux committed:
docs(Recording): update docs to match implementation

vigoux committed:
style(Recording): rename handler to match suggestions

vigoux committed:
fix(RecordingLeave): send autocommand earlier

This makes the autocommand fire just before setting reg_recorded to
reg_recording, this way we clearly show that we are actually just before
actually quitting the recording mode.

vigoux commented:
I should have addressed all the comments, so I think I just need @justinmk 's approval on this and we'll be good to go !

zeertzjq commented:
There are still some references to the old `Q` in docs:

- `eval.txt` line 6884
- `vim_diff.txt` line 358
- `repeat.txt` line 106 (this should be changed to `gQ`)
- `index.txt` line 342
- `quickref.txt` line 999
- `options.txt` line 953 (I have mentioned this above)

zeertzjq commented:
> There are still some references to the old `Q` in docs:
> 
>     * `eval.txt` line 6884
> 
>     * `vim_diff.txt` line 358
> 
>     * `repeat.txt` line 106 (this should be changed to `gQ`)
> 
>     * `index.txt` line 342
> 
>     * `quickref.txt` line 999
> 
>     * `options.txt` line 953 (I have mentioned this above)
> 
> 
> There are also a lot of oldtests failing because they use `Q`.

I think the oldtest failures can be solved by mapping `Q` to `gQ` in `runtest.vim`.

AckslD commented:
@vigoux @zeertzjq Is there anyway I can help out with getting this ready? For example updating the docs and tests as you pointed out @zeertzjq or something else. Really looking forward to this so I'm happy to help out if possible :)

gpanders commented:
@AckslD @vigoux is pretty busy right now with other commitments, so feel free to take over this PR. At this point it should just need test and doc updates.

Be sure to throw a reference/link to this PR when you create your new one.

vigoux commented:
Hey sorry for the delay, you can of course take this PR over, and you can drop me an email whenever you need any help / guidance, and I'll try my best to help you out.

Thanks for taking that over BTW !

AckslD commented:
Hi @gpanders @vigoux, no problem, I'll try to take a look at this and update test and docs and get back to you when done or if I have questions :)

zeertzjq commented on runtime/doc/autocmd.txt:
```diff
@@ -815,6 +815,14 @@ RemoteReply			When a reply from a Vim that functions as
+				The pattern is the current file name, and
+				|reg_recording()| is the current register that
+				is used.
+							*RecordinLeave*
```
I just found a typo here.

AckslD replied:
fixed in #16684

AckslD commented:
I just opened #16684 where I so far fixed the things in the docs that @zeertzjq pointed out.

@zeertzjq regarding the old tests, are there ever new tests added there? Just wondering if it would be fine to just map `Q` to `gQ` as you suggested.

gpanders commented:
Moved to #16684.
